Plugins/FeedWriter: Add support for bucket RSS2/Atom feeds.

nlewis2142 opened the pull request:
Closes #174

Complete functionality is added. Still need designer to decide if RSS button is correct and add Atom button (@brandonrosage).

Binds to the following URLs: `<account>/bucket/<name>/(rss|atom)`

Injects RSS button into bucket index: `<account>/bucket/<name>(/<action>)`

Commit:
Plugins/FeedWriter: Basic RSS functionality implemented

Commit:
Plugins/FeedWriter: Code cleanup

Commit:
Plugins/FeedWriter: More code cleanup and minor bugfix

Commit:
Plugins/FeedWriter: Added RSS button through JQuery injection.

Commit:
Plugins/FeedWriter: Minor change to the way dates are calculated

Commit:
Plugins/FeedWriter: Move time conversion inside RSS class.

Commit:
Plugins/FeedWriter: Added basic atom feed support

Commit:
Plugins/FeedWriter: More code cleanup and readability.

Commit:
Plugins/FeedWriter: Atom fully operation, needs a button from design.

Commit:
Plugins/FeedWriter: Removed a check that wasn't doing anything

nlewis2142 commented:
Ready for merge.

69mb commented on plugins/feedwriter/classes/controller/feedwriter.php:
```diff
@@ -0,0 +1,307 @@
+    }
+}
+
+class Feedwriter_Rss
```
Move this class to its own file under plugins/feedwriter/classes/feedwriter/rss.php

nlewis2142 replied:
@69mb Already attempted this, wasn't sure how to get Kohana to find the file as the usual include/require didn't work. What would be the proper way to include a plugin library?

69mb replied:
Just tried this on my local and it works ok.

nlewis2142 replied:
Verified, just got it working first try. Not sure how this was failing before.

69mb commented on plugins/feedwriter/classes/controller/feedwriter.php:
```diff
@@ -0,0 +1,307 @@
+    }
+}
+
+class Feedwriter_Atom
```
Same as with Feedwriter_RSS above, move this to plugins/feedwriter/classes/feedwriter/atom.php

69mb commented on plugins/feedwriter/classes/controller/feedwriter.php:
```diff
@@ -0,0 +1,307 @@
+        $this->meta = array(
+            'title'         => ($title) ? $title : "RSS2 Feed",
+            'link'          => ($link) ? $link :
+                URL::site($_SERVER["REQUEST_URI"], true),
```
use URL::base(TRUE, TRUE) instead of referencing $_SERVER["REQUEST_URI"]

nlewis2142 replied:
@69mb That does not give the desired input.

That outputs http://swiftriver.beta/
It should be outputting http://swiftriver.beta/admin/bucket/my-bucket/rss

URL::site is an extension of URL::base but includes subdirectories.

69mb replied:
Use $this->request->url then?

nlewis2142 replied:
Looked it up on Kohana forums. Correct usage should be

URL::site(Kohana_Request::detect_uri(), true)

Giving this a try.

69mb commented on plugins/feedwriter/init.php:
```diff
@@ -0,0 +1,39 @@
+
+    public static function Inject()
+    {
+        echo '<script type="text/javascript">$(document).ready(function(){$(".page-title.bucket-title.cf div h1 span").after(\'<span class="rss-feed" style="margin-left: 20px; display: none;"><a href="rss" target="_blank"><img src="/themes/default/media/img/channel-rss.gif" alt="RSS" title="RSS" height="12" style="margin-bottom: 4px;" /></a></span>\');$("span.rss-feed").fadeIn()})</script>';
```
I would suggest for separation that the javascript goes into its own file, either:
- Put the javascript into plugins/rss/media/js/icon.js and in this line echo Html::script("plugins/rss/media/js/icon.js")
- Or put the script tag in its own view in plugins/rss/views/icon.php and echo the view at this line View::factory('plugins/rss/views/icon.php')

nlewis2142 replied:
@69mb Changing that now.

69mb commented:
Good stuff. I think the following would also complete this feature:
- Add the alternative meta links to the header of the bucket page.
- Add either an etag/modified header this way crawlers can tell if there has been an update without doing a full request.
- Permissions. A crawler cannot access the url on a deployment that does not allow anonymous access / for private buckets. Either we do not provide a feed in such cases (easy way out) or have some token in the url that will allow access(make use of auth_tokens?). Similar to having the logged in user's api_key in the url will allow access but the api_key is sensitive and should not be shared this way...

@brandonrosage help with the icon?

69mb commented:
Seems there is also a content type issue with atom --> https://dl.dropbox.com/u/48291217/Screenshots/kazk.png

rss is ok --> https://dl.dropbox.com/u/48291217/Screenshots/z7uu.png

Commit:
Plugins/FeedWriter: Move JS bulk into media/js.

Commit:
Plugins/FeedWriter: Fix content-type in Atom

nlewis2142 commented:
@69mb Can you check atom content again?

69mb commented:
+1 Atom is now ok

Commit:
Plugins/FeedWriter: Split libs and switch to Kohana's URI-Lib

nlewis2142 commented:
@69mb Removed the $_SERVER["REQUEST_URI"] and moved the RSS+Atom libs to their own folder/document. Have I missed anything?

Edit: Doomed to miss something, just added the "die on direct access" and header info in latest commit.

Commit:
Plugins/FeedWriter: Missed the header info and die at the top:

ekala commented on plugins/feedwriter/classes/libs/atom.php:
```diff
@@ -0,0 +1,131 @@
+        );
+    }
+
+    public function setTitle($title)
```
A note on the coding style convention: we usually use lowercase and an underscore( `_`) to separate words

nlewis2142 replied:
as in `set_link()`?

ekala replied:
Exactly.

Commit:
Plugins/FeedWriter: Add meta injection

Commit:
Plugins/FeedWriter: update coding style on function names

nlewis2142 commented:
@ekala Changed. Anything else?

Commit:
Plugins/FeedWriter: Added ETag support

ekala commented:
Getting this error: http://cl.ly/1T0Y0t1b0G2b1t1J1j1T

nlewis2142 commented:
@ekala Not part of this branch/pull as far as I know. I did not add any `Kohana::log()` functions.

My Kohana log does not generate any errors on page load either. It may be an error presented in one of the recent changes to the master tree.

ekala commented:
Sorry man. I uploaded the wrong screenie :fist: Here's the correct one. http://cl.ly/3s2V2l423w3C3t380S0a

nlewis2142 commented:
@ekala I'm still not getting that error. Can you view source and paste that here? There's likely an entity that isn't showing up in that error.

Edit: It stopped rendering after the `<item><link>` element which means the error was in the description/content of the droplet.

Commit:
Plugins/FeedWriter: Giving htmlspecialchars a try w/ UTF-8
